docs/pnpm.md

```diff
@@ -0,0 +1,282 @@
+If the version is left out of the package name, the patch will be applied to every version of the npm package.
+
+`patch_args` defaults to `-p0`, but `-p1` will usually be needed for patches generated by git.
+
+In case multiple entries in `patches` match, the list of patches are additive.
+(More specific matches are appended to previous matches.)
+However if multiple entries in `patch_args` match, then the more specific name matches take precedence.
+
+For example,
+
+```starlark
+npm_translate_lock (
+    ...
+    patches = {
+        "@foo/bar": ["//:patches/foo+bar.patch"],
+        "fum@0.0.1": ["//:patches/fum@0.0.1.patch"],
+    },
+    patch_args = {
+        "*": ["-p1"]
+        "@foo/bar": ["-p0"]
+        "fum@0.0.1": ["-p2"]
+    },
+)
+```
+
+### Lifecycles
+
+npm packages have "lifecycle scripts" such as `postinstall` which are documented here:
+<https://docs.npmjs.com/cli/v9/using-npm/scripts#life-cycle-scripts>
+
+We refer to these as "lifecycle hooks".
+You can disable this feature completely by setting `lifecycle_hooks = False` in `npm_translate_lock`.
+
+Because rules_js models the execution of these hooks as build actions, rather than repository rules,
+the result can be stored in the remote cache and shared between developers.
+Typically these actions are not run in Bazel's action sandbox because of the overhead of setting up and tearing down the sandboxes.
+
+> You can set `lifecycle_hooks_no_sandbox = False` in `npm_translate_lock` to change this default.
+
+In addition to sandboxing, Bazel supports other `execution_requirements` for actions,
+in the attribute of <https://bazel.build/rules/lib/actions#run>.
+You can have control over these using the `lifecycle_hooks_execution_requirements` attribute of `npm_translate_lock`.
+
+Some hooks may fail to run under rules_js, and you don't care to run them.
+You can use the `lifecycle_hooks_exclude` attribute of `npm_translate_lock` to turn them off for a package,
+which is equivalent to setting the `lifecycle_hooks` to an empty list for that package.
+
+You can set environment variables for hook build actions using the `lifecycle_hooks_envs` attribute of `npm_translate_lock`.
+
+In case there are multiple matches, some attributes are additive.
+(More specific matches are appended to previous matches.)
+Other attributes have specificity: the most specific match wins and the others are ignored.
+
+| attribute                              | behavior    |
+| -------------------------------------- | ----------- |
+| lifecycle_hooks                        | specificity |
+| lifecycle_hooks_envs                   | additive    |
+| lifecycle_hooks_execution_requirements | specificity |
+
+Here's a complete example of managing lifecycles:
+
+```starlark
+npm_translate_lock(
+    ...
+    lifecycle_hooks = {
+        # These three values are the default if lifecycle_hooks was absent
+        "*": [
+            "install",
```

Comment:
order matters here and the order is typically: preinstall, install, postinstall

Reply:
oh good to comment then